### PR TITLE
Clarify `ReplicaConsistency` vs commits

### PR DESCRIPTION
**Pull request by dagguh**
This is how I understood the SPI. This draft is just a place for discussion and clarification.

### CODE REVIEW

File: src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java [hunk 1]
```diff
@@ -57,14 +57,7 @@ public int getTransactionIsolation() throws SQLException {
     }
 
     public void setAutoCommit(Boolean autoCommit) throws SQLException {
-        final boolean autoCommitBefore = getAutoCommit();
-        if (autoCommitBefore != autoCommit) {
-            preCommit(autoCommitBefore);
-        }
         parameters.setAutoCommit(state::getConnection, autoCommit);
-        if (autoCommitBefore != getAutoCommit()) {
-            recordCommit(autoCommitBefore);
```
Comment (dagguh) [inline src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java:66]: Why did `ReplicaConsistency` need to know about `setAutoCommit`? Can this call cause an immediate commit?
I thought that the next `executeUpdate` could do that implicitly, but nothing happens in `setAutoCommit` itself.

Comment (dagguh) [reply src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java:66]: I added the integration tests. One fails when introduced initially and goes green after removing auto-commit handling.

File: src/test/java/com/atlassian/db/replica/api/TestConsistency.java [hunk 2]
```diff
@@ -137,22 +136,6 @@ public void shouldNotRefreshAfterSelectForUpdate() throws SQLException {
         verify(consistency, never()).write(any());
     }
 
-    @Test
-    public void shouldSetAutoCommitRefreshWhenInTransaction() throws SQLException {
```
Comment (afaruga-atlassian) [inline src/test/java/com/atlassian/db/replica/api/TestConsistency.java:141]: If  test is deleted, shouldn't a new one be created for reusing transactions 🤔

Comment (dagguh) [reply src/test/java/com/atlassian/db/replica/api/TestConsistency.java:141]: There are two new tests in `DualConnectionIT`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -24,6 +24,10 @@ Dropping a requirement of a major version of a dependency is a new contract.
 ## [Unreleased]
 [Unreleased]: https://github.com/atlassian-labs/db-replica/compare/release-1.3.0...master
 
+### Added
+- Call `ReplicaConsistency.preCommit` only for explicit commits.
```
Comment (wyrzyk) [inline CHANGELOG.md:28]: It was working this way from the beginning.

File: src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java (hunk 1 above)
Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java:66]: > Why did ReplicaConsistency need to know about setAutoCommit? Can this call cause an immediate commit?

Yes. It's a part of the connection API. 
```
   * <B>NOTE:</B>  If this method is called during a transaction and the
     * auto-commit mode is changed, the transaction is committed.  If
     * <code>setAutoCommit</code> is called and the auto-commit mode is
     * not changed, the call is a no-op.
```

**Comment by dagguh** [comment]
Thanks for clarifying

**Pull request closed without merging**